File: abb/abb.c
#include "abb.h"
void inicia_arvore (t_abb *abb) {
    abb->raiz = NULL;
}

int esta_vazia (t_abb *abb) {
    return abb->raiz == NULL;
}

void insere_rec (t_no *novo, t_no *atual) {
    if (novo->info > atual->info) { //ir para direita
        if (atual->direita == NULL) 
            atual->direita = novo;
        else
            insere_rec (novo, atual->direita);
    }
    else { //ir para esquerda
        if (atual->esquerda == NULL)
            atual->esquerda = novo;
        else
            insere_rec (novo, atual->esquerda);
    }
}
void insere (int i, t_abb *abb) {
    t_no *novo = constroi_no(i);
    if (esta_vazia(abb)) {
        abb->raiz = novo;
    }
    else {
        insere_rec (novo, abb->raiz);
    }
}

void print_em_ordem_rec (t_no * atual) {
    if (atual->esquerda != NULL) // navega para esquerda
    print_em_ordem_rec (atual->esquerda);
    printf ("%d ", atual->info); // faz o processamento da informação atual
    if (atual->direita != NULL)  // navega para direita
    print_em_ordem_rec (atual->direita);
}
void print_em_ordem (t_abb *abb) {
    if (esta_vazia(abb))
        printf ("arvore vazia\n");
    else 
        print_em_ordem_rec (abb->raiz);
}

int conta_nos_rec (t_no *atual) {
    if (atual == NULL) return 0;
    return 1 + conta_nos_rec(atual->esquerda) + conta_nos_rec(atual->direita);
}
int conta_nos (t_abb *abb) {
    if (esta_vazia(abb)) return 0;
    return conta_nos_rec (abb->raiz);
}
int soma_valores_rec (t_no *atual) {
    if (atual == NULL) return 0;
    return atual->info + soma_valores_rec(atual->esquerda) + soma_valores_rec(atual->direita);
}
int soma_valores (t_abb *abb){
    if (esta_vazia(abb)) return 0;
    return soma_valores_rec (abb->raiz);
}

File: abb/ABB.java
public class ABB {
    private No raiz;
    //construtor é o padrão, isto é, raiz = null
    //não temos get e set porque manipular a raiz é de interesse só da própria classe

    public boolean estaVazia () {
        return raiz == null;
    }
    public void insere (int info) {
        No novo = new No(info);
        if (estaVazia()) raiz = novo;
        else insereRec (novo, raiz);
    }
    private void insereRec (No novo, No atual) {
        if (novo.getInfo() > atual.getInfo()) {//vai para direita
            if (atual.getDireito() == null)
                atual.setDireito(novo);
            else 
                insereRec(novo, atual.getDireito());
        }
        else {//vai para esquerda
            if (atual.getEsquerdo() == null) 
                atual.setEsquerdo(novo);
            else 
                insereRec(novo, atual.getEsquerdo());
        }
    }
    @Override
    public String toString () {
        if (estaVazia()) return "arvore vazia\n";
        return toStringRec1(raiz);
    }
    private String toStringRec1 (No atual) {
        String s = "";
        if (atual.getEsquerdo() != null) s += toStringRec1(atual.getEsquerdo());
        s += atual.getInfo() + " ";
        if (atual.getDireito() != null) s += toStringRec1(atual.getDireito());
        return s;
    }
    private String toStringRec2 (No atual) {
        String s = "";
        if (atual != null)
            s += toStringRec2(atual.getEsquerdo()) + 
                 atual.getInfo() + " " + 
                 toStringRec2(atual.getDireito());
        return s; 
    }
    
    public int contaNos() {
        return contaNosRec(raiz);
    }
    private int contaNosRec(No atual) {
        if (atual == null) return 0;
        return 1 + contaNosRec(atual.getEsquerdo()) + contaNosRec(atual.getDireito());
    }
    public int somaValores() {
        return somaValoresRec(raiz);
    }
    private int somaValoresRec(No atual) {
        if (atual == null) return 0;
        return atual.getInfo() + 
               somaValoresRec(atual.getEsquerdo()) + somaValoresRec(atual.getDireito());
    }
}
